oeps/oep-0052-arch-event-bus-technology.rst

```diff
@@ -0,0 +1,264 @@
+
+Also see a useful and biased `comparison of Apache Kafka vs Vendors`_ by Kai Waehner (of Confluent), comparing various providers and distributions of Kafka and related or competitive services.
+
+.. _Amazon MSK: https://aws.amazon.com/msk/
+.. _Confluent Platform: https://www.confluent.io/product/confluent-platform
+.. _comparison of Apache Kafka vs Vendors: https://www.kai-waehner.de/blog/2021/04/20/comparison-open-source-apache-kafka-vs-confluent-cloudera-red-hat-amazon-msk-cloud/
+
+Kafka Highlights
+~~~~~~~~~~~~~~~~
+
+Pros
+^^^^
+
+* Battle-tested, widely adopted, big community, lots of documentation and answers.
+* Amazon MSK (AWS service) provides hosted path of least resistance.
+* `New Relic integration with Amazon MSK`_ (useful to edX.org).
+
+Cons
+^^^^
+
+* Many open questions about add-ons required for developers and operators.
+* Complex to manage, including likely manual scaling.
+
+.. _New Relic integration with Amazon MSK: https://docs.newrelic.com/docs/integrations/amazon-integrations/aws-integrations-list/aws-managed-kafka-msk-integration/
+
+Consequences
+------------
+
+* Operators will need to deploy and manage the selected infrastructure, which is likely to be complex. If Apache Kafka is selected, there are likely to be a set of auxiliary parts to provide all required functionality for our message bus.
+* Education will be required for both developers and operators regarding best practices for each role.
+* Code to interact with Kafka and its libraries will be added to core services.
+* At least one initial use case must be completed. One potential candidate is the grade change event in the LMS, and its use by the Credentials service.
+* Once we have a message bus, we can investigate other potential use cases:
+
+  * Course/program update propagation.
+  * Feed into xAPI/Caliper capabilities.
+  * New services and features can be built fully de-coupled from the core application.
+
+Rejected Alternatives
+---------------------
+
+Apache Pulsar
+~~~~~~~~~~~~~
+
+Although rejected to start, `Apache Pulsar`_ remains an option if solving with Kafka turns out to be overly burdensome for developers or operators.
+
+Pros
+^^^^
+
+* Ease of scalability (built-in, according to docs).
+* Ease of data retention capabilities.
+* Additional built-in pub/sub features (built-in, according to docs).
+
+Cons
+^^^^
+
+* Requires 3rd party hosting or larger upfront investment in self-hosted (kubernetes).
+* Less mature (but growing) community, little documentation, and few answers.
+
+Note: Read an interesting (Kafka/Confluent) biased article exploring `comparisons and myths of Kafka vs Pulsar`_.
+
+.. _Apache Pulsar: https://pulsar.apache.org/
+.. _comparisons and myths of Kafka vs Pulsar: https://dzone.com/articles/pulsar-vs-kafka-comparison-and-myths-explored
+
+Redis
+~~~~~
+
+Pros
+^^^^
+
+* Already part of Open edX platform
+
+Cons
+^^^^
+
+* Can lose acked data, even if RAM backed up with an append-only file (AOF).
```

Comment:
In practice, exactly-once processing is very difficult to achieve with all brokers, including Kafka. This argument should be balanced with the difficulty of deploying an exactly-once data processing broker.

Reply:
> In general, the hosting and deployment story for each Open edX deployment makes this OEP that much more complex. In the case of Amazon’s ElastiCache for Redis (which we use), AOF is not an option for multi-AZ, and they state that multi-AZ has better resiliency. However, it’s not a great story for disaster recovery, which means some homegrown solution of backing up and being able to restore all events. Instead, I think we are going to need a solution with persistence.

The fact that ElastiCache for Redis does not come with persistence has come up before (for the replacement of Rabbitmq) and it baffles me. It means that on AWS Redis can only be used as a cache layer, which drastically limits its capabilities. Thus, Redis should not be deployed with ElastiCache. But why limit yourself to AWS-offered options? Redis Enterprise offers AWS-hosted solutions that come with persistence out of the box: https://redis.com/redis-enterprise-cloud/pricing/ I am sure there are other Redis providers out there that offer a better storage story.

> Another con that is missing here is that because Redis is in-memory, it is limited by RAM. I also don’t think it has a good compaction story. This seems like a hard limit that again would be very costly to work around, like trying to manage and balance across multiple Redis clusters, or any other custom solution we might come up with.

Do we have an estimate for the amount of data (/user/day) that we would need to store? I know that deploying a full blown event-driven architecture requires storing all events forever, but surely this is intractable for most Open edX system administrators -- even when stored on disk. In order to dismiss RAM-based storage, we need an estimate for the amount of stored data.